backend/middleware/auth.js

```diff
@@ -1,13 +1,69 @@
-async function withAnyAuth(req, res, next) {
-  next();
+// Copyright Schulich Racing, FSAE
+// Written by Justin Tijunelis, Arham Humayun
+"use strict";
+/*
+* Calls the next function if the auth is the required one or higher
+*/
+
+const call = require("../utilities/call");
+const permissionValues = {
+  Pending: 1,
+  Guest: 2,
+  Member: 3,
+  Lead: 4,
+  Admin: 5,
 }
 
-async function withAdminAuth(req, res, next) {
-  next();
+async function hasAuth(req, res, next, required_perms) {
+
+  // Parse header for cookies
+  const rawCookies = req.headers.cookie.split('; ');
+  const parsedCookies = {};
+  rawCookies.forEach(rawCookie=>{
+  const parsedCookie = rawCookie.split('=');
+    parsedCookies[parsedCookie[0]] = parsedCookie[1];
+  });
+
+  // Verify if Authorization cookie exists
+  if (!parsedCookies.Authorization) {
+    res.status(401).send({ error: "Authorization cookie not found." }).end()
+  }
+
+  // Call DB MS route to verify cookie
+  const path = process.env.DATABASE_MS_ROUTE + "/user/role/" + parsedCookies.Authorization;
+  const authRes = await call(path, "GET");
+
+  // Handle response
+  if (authRes.body && authRes.body.success) {
+    if (permissionValues[authRes.body.role] >= permissionValues[required_perms]) {
+      next();
+    } else {
+    res.status(401).send({ error: "Permission denied." }).end()
+    }
+  } else {
+    res.status(401).send({ error: "Authorizing cookie failed." }).end()
+  }
+}
+
+async function withPendingAuth(req, res, next) {
+  await hasAuth(req, res, next, "Pending")
 }
 
-async function withCaptainAuth(req, res, next) {
-  next();
+async function withGuestAuth(req, res, next) {
+  await hasAuth(req, res, next, "Guest")
 }
 
-module.exports = { withAnyAuth, withAdminAuth, withCaptainAuth };
+async function withMemberAuth(req, res, next) {
+  await hasAuth(req, res, next, "Member")
+}
+
+async function withLeadAuth(req, res, next) {
+  await hasAuth(req, res, next, "Lead")
+}
+
+async function withAdminAuth(req, res, next) {
+  await hasAuth(req, res, next, "Admin")
+}
+
```

Comment:
Maybe we should make a withAnyAuth? Bc renew, change password, etc, should be able to handle that.

Reply:
We can use withPendingAuth as a withAnyAuth since the lowest permission role a user will have will be Pending.

Reply:
How would we resolve that though? withAnyAuth = "Pending" || "Lead" || etc, where withPendingAuth = "Pending". Additionally, not sure if the semantics make sense, how could an admin be pending?

Reply:
withPendingAuth would also pass if a user is an Admin since we check permissions in a tiered way. withPendingAuth doesn't just pass if the role is "Pending", it'll pass if the users role is > "Pending", so all roles above "Pending" will pass. We could create a withAnyAuth function for semantics' sake, but it will be an identical function to withPendingAuth.

Reply:
Okay gotcha, will wait for Jeremy's input

Reply:
Looks good to me. The function names could be changed in a future commit for clarity but i agree that the functionality will be the same.